### PR TITLE
fix(lsp): accept file URIs without a hostname

### PR DESCRIPTION
**Pull request by donbex**
RFC 8089, which defines the file URI scheme, also [allows URIs without a hostname](https://datatracker.ietf.org/doc/html/rfc8089#section-2), i.e. of the form `file:/path/to/file` instead of `file:///path/to/file`. These are returned by some language servers (e.g. [vdmj](https://github.com/nickbattle/vdmj/issues/54)), causing neovim to throw an error after initialisation. Since they accepted by other LSP implementations, such as the ones in VSCode and Eclipse's lsp4e, it is reasonable for neovim to accept them as well.

### CODE REVIEW

**Comment by mjlbach** [comment]
Seems fine to me, vscode also can ignore the authority as well.

https://github.com/microsoft/vscode-uri/blob/9f8d0693d198c578f91e17ba3b93be144840939c/src/uri.ts#L27-L32

@mfussenegger?

File: runtime/lua/vim/uri.lua [hunk 1]
```diff
@@ -52,7 +52,7 @@ end
 
 --@private
 local function is_windows_file_uri(uri)
-  return uri:match('^file:///[a-zA-Z]:') ~= nil
+  return uri:match('^file:/+[a-zA-Z]:') ~= nil
```
Comment (mjlbach) [inline runtime/lua/vim/uri.lua:55]: Hmm would it be better to optionally match on [//] ?

Comment (donbex) [reply runtime/lua/vim/uri.lua:55]: If I understand your suggestion correctly, that's actually what I wanted to do at first. Unfortunately Lua's pattern matching system is much less powerful than regular expressions and it does not allow optional matches (besides repetition).

**Comment by donbex** [comment]
I am not convinced by the correctness of the failing tests in `lsp_spec.lua`, e.g.https://github.com/neovim/neovim/blob/4547137aaff32b20172870a549d3a28a3c7adf1c/test/functional/plugin/lsp_spec.lua#L1532-L1557
According to the file URI specification, the URI `file://fake/uri` would point to the path `/uri` on hostname `fake`, not to the path `fake/uri` on the local machine.

Is there a specific reason why this was chosen? For example, are there language servers that return such malformed URIs?

**Comment by mfussenegger** [comment]
> Seems fine to me, vscode also can ignore the authority as well.
> 
> https://github.com/microsoft/vscode-uri/blob/9f8d0693d198c578f91e17ba3b93be144840939c/src/uri.ts#L27-L32
> 
> @mfussenegger?

Yes I think we should make the authority optional as well - by pulling in the changes from https://github.com/neovim/neovim/pull/14959

**Comment by donbex** [comment]
What is the preferred workflow for that, @mfussenegger? Should I cherry pick those commits, rebase my OR on top of that branch, or something else entirely?

**Comment by mfussenegger** [comment]
> Should I cherry pick those commits, rebase my OR on top of that branch

That sounds good :+1:

**Comment by donbex** [comment]
I have cherry picked the commit from https://github.com/neovim/neovim/pull/14959 and rebased on top of the current master. :+1:

File: runtime/lua/vim/uri.lua (hunk 1 above)
Comment (mjlbach) [reply runtime/lua/vim/uri.lua:55]: Yes, we could vim's built-in regex engine, but yes, this is probably fine. I just think it's a little weird we could match against  `file://////////`

**Review by mfussenegger — APPROVED** [review]

**Review by mjlbach — APPROVED** [review]

**Comment by tapayne88** [comment]
Hi all, is this change a candidate to be included in the next patch release of v0.5?

**Comment by mjlbach** [comment]
Not until we fix some related URI issues that this caused on windows

**Comment by tapayne88** [comment]
cool, thanks @mjlbach. Do you know if this type of change would automatically be rolled into the next patch (when all issues are resolved) or should I keep watching and pestering?

**Pull request merged**